File: mobile/src/services/wifiService.ts
// Tipos para el servicio WiFi
export interface WiFiNetwork {
  ssid: string;
  rssi: number;
  security: 'Open' | 'WPA' | 'WPA2' | 'WPA3' | 'WEP';
  frequency: number;
  isConnected?: boolean;
}

export interface WiFiCredentials {
  ssid: string;
  password: string;
}

export interface WiFiConnectionResult {
  success: boolean;
  message: string;
  connectedNetwork?: WiFiNetwork;
}

class WiFiService {
  private isScanning: boolean = false;
  private currentNetwork: WiFiNetwork | null = null;

  /**
   * Escanear redes WiFi disponibles
   */
  async scanNetworks(): Promise<WiFiNetwork[]> {
    if (this.isScanning) {
      console.warn('WiFi scan already in progress');
      return [];
    }

    try {
      this.isScanning = true;

      // Simular escaneo de redes WiFi
      const mockNetworks: WiFiNetwork[] = [
        {
          ssid: 'Mi Casa WiFi',
          rssi: -30,
          security: 'WPA2',
          frequency: 2400,
        },
        {
          ssid: 'Vecino WiFi',
          rssi: -60,
          security: 'WPA3',
          frequency: 5000,
        },
        {
          ssid: 'Café Internet',
          rssi: -70,
          security: 'Open',
          frequency: 2400,
        },
        {
          ssid: 'Oficina WiFi',
          rssi: -45,
          security: 'WPA2',
          frequency: 2400,
        },
        {
          ssid: 'NEBU-Setup',
          rssi: -35,
          security: 'Open',
          frequency: 2400,
        },
      ];

      // Simular tiempo de escaneo
      await new Promise(resolve => setTimeout(resolve, 2000));

      this.isScanning = false;
      
      // Ordenar por intensidad de señal (RSSI más alto = mejor señal)
      return mockNetworks.sort((a, b) => b.rssi - a.rssi);
    } catch (error) {
      this.isScanning = false;
      console.error('Error scanning WiFi networks:', error);
      throw error;
    }
  }

  /**
   * Conectar a una red WiFi
   */
  async connectToNetwork(credentials: WiFiCredentials): Promise<WiFiConnectionResult> {
    try {
      console.log(`Attempting to connect to WiFi: ${credentials.ssid}`);

      // Simular proceso de conexión
      await new Promise(resolve => setTimeout(resolve, 3000));

      // Simular diferentes resultados basados en las credenciales
      const connectionSuccess = this.simulateConnectionResult(credentials);

      if (connectionSuccess) {
        const connectedNetwork: WiFiNetwork = {
          ssid: credentials.ssid,
          rssi: -40,
          security: 'WPA2',
          frequency: 2400,
          isConnected: true,
        };

        this.currentNetwork = connectedNetwork;

        return {
          success: true,
          message: 'Successfully connected to WiFi network',
          connectedNetwork,
        };
      } else {
        return {
          success: false,
          message: 'Failed to connect to WiFi network. Please check your credentials.',
        };
      }
    } catch (error) {
      console.error('Error connecting to WiFi:', error);
      return {
        success: false,
        message: 'An error occurred while connecting to WiFi',
      };
    }
  }

  /**
   * Verificar conexión a internet
   */
  async checkInternetConnection(): Promise<boolean> {
    try {
      // Simular verificación de conexión a internet
      await new Promise(resolve => setTimeout(resolve, 1000));
      
      // Simular 90% de éxito en la conexión
      return Math.random() > 0.1;
    } catch (error) {
      console.error('Error checking internet connection:', error);
      return false;
    }
  }

  /**
   * Obtener información de la red actual
   */
  getCurrentNetwork(): WiFiNetwork | null {
    return this.currentNetwork;
  }

  /**
   * Desconectar de la red actual
   */
  async disconnect(): Promise<void> {
    try {
      console.log('Disconnecting from current WiFi network');
      
      // Simular desconexión
      await new Promise(resolve => setTimeout(resolve, 1000));
      
      this.currentNetwork = null;
      console.log('Disconnected from WiFi network');
    } catch (error) {
      console.error('Error disconnecting from WiFi:', error);
      throw error;
    }
  }

  /**
   * Validar credenciales WiFi
   */
  validateCredentials(credentials: WiFiCredentials): { isValid: boolean; message: string } {
    if (!credentials.ssid || credentials.ssid.trim().length === 0) {
      return {
        isValid: false,
        message: 'WiFi network name (SSID) is required',
      };
    }

    if (credentials.ssid.length < 2) {
      return {
        isValid: false,
        message: 'WiFi network name must be at least 2 characters long',
      };
    }

    if (credentials.ssid.length > 32) {
      return {
        isValid: false,
        message: 'WiFi network name cannot exceed 32 characters',
      };
    }

    // Validar contraseña si no es una red abierta
    if (credentials.password && credentials.password.length > 0) {
      if (credentials.password.length < 8) {
        return {
          isValid: false,
          message: 'WiFi password must be at least 8 characters long',
        };
      }

      if (credentials.password.length > 63) {
        return {
          isValid: false,
          message: 'WiFi password cannot exceed 63 characters',
        };
      }
    }

    return {
      isValid: true,
      message: 'Credentials are valid',
    };
  }

  /**
   * Obtener nivel de señal como texto
   */
  getSignalLevel(rssi: number): string {
    if (rssi >= -30) return 'Excellent';
    if (rssi >= -50) return 'Good';
    if (rssi >= -70) return 'Fair';
    if (rssi >= -80) return 'Weak';
    return 'Very Weak';
  }

  /**
   * Obtener ícono de seguridad para la red
   */
  getSecurityIcon(security: string): string {
    switch (security) {
      case 'Open':
        return '🔓';
      case 'WEP':
        return '🔒';
      case 'WPA':
        return '🔐';
      case 'WPA2':
        return '🔐';
      case 'WPA3':
        return '🔐';
      default:
        return '❓';
    }
  }

  /**
   * Simular resultado de conexión (para testing)
   */
  private simulateConnectionResult(credentials: WiFiCredentials): boolean {
    // Simular diferentes escenarios
    if (credentials.ssid.toLowerCase().includes('casa') || 
        credentials.ssid.toLowerCase().includes('home')) {
      return Math.random() > 0.1; // 90% success for home networks
    }
    
    if (credentials.ssid.toLowerCase().includes('oficina') || 
        credentials.ssid.toLowerCase().includes('office')) {
      return Math.random() > 0.2; // 80% success for office networks
    }
    
    if (credentials.ssid.toLowerCase().includes('cafe') || 
        credentials.ssid.toLowerCase().includes('public')) {
      return Math.random() > 0.3; // 70% success for public networks
    }
    
    // Redes con contraseñas muy simples fallan más
    if (credentials.password && credentials.password.length < 8) {
      return Math.random() > 0.5; // 50% success for weak passwords
    }
    
    // Por defecto, 85% de éxito
    return Math.random() > 0.15;
  }

  /**
   * Obtener estado del escaneo
   */
  getScanStatus(): { isScanning: boolean } {
    return {
      isScanning: this.isScanning,
    };
  }
}

// Instancia singleton del servicio
export const wifiService = new WiFiService();
export default wifiService;

File: lib/data/services/wifi_service.dart
import 'dart:async';
import 'package:logger/logger.dart';
import 'package:permission_handler/permission_handler.dart';
import 'package:wifi_scan/wifi_scan.dart';

/// Red WiFi
class WiFiNetwork {
  const WiFiNetwork({
    required this.ssid,
    required this.rssi,
    required this.security,
    required this.frequency,
    this.isConnected = false,
  });

  factory WiFiNetwork.fromJson(Map<String, dynamic> json) => WiFiNetwork(
    ssid: json['ssid'] as String,
    rssi: json['rssi'] as int,
    security: json['security'] as String,
    frequency: json['frequency'] as int,
    isConnected: json['isConnected'] as bool? ?? false,
  );
  final String ssid;
  final int rssi;
  final String security; // 'Open' | 'WPA' | 'WPA2' | 'WPA3' | 'WEP'
  final int frequency;
  final bool isConnected;

  Map<String, dynamic> toJson() => {
    'ssid': ssid,
    'rssi': rssi,
    'security': security,
    'frequency': frequency,
    'isConnected': isConnected,
  };
}

/// Credenciales WiFi
class WiFiCredentials {
  const WiFiCredentials({required this.ssid, required this.password});

  factory WiFiCredentials.fromJson(Map<String, dynamic> json) =>
      WiFiCredentials(
        ssid: json['ssid'] as String,
        password: json['password'] as String,
      );
  final String ssid;
  final String password;

  Map<String, dynamic> toJson() => {'ssid': ssid, 'password': password};
}

/// Resultado de conexión WiFi
class WiFiConnectionResult {
  const WiFiConnectionResult({
    required this.success,
    required this.message,
    this.connectedNetwork,
  });

  factory WiFiConnectionResult.fromJson(Map<String, dynamic> json) =>
      WiFiConnectionResult(
        success: json['success'] as bool,
        message: json['message'] as String,
        connectedNetwork: json['connectedNetwork'] != null
            ? WiFiNetwork.fromJson(
                json['connectedNetwork'] as Map<String, dynamic>,
              )
            : null,
      );
  final bool success;
  final String message;
  final WiFiNetwork? connectedNetwork;

  Map<String, dynamic> toJson() => {
    'success': success,
    'message': message,
    'connectedNetwork': connectedNetwork?.toJson(),
  };
}

/// Servicio WiFi
class WiFiService {
  WiFiService({required Logger logger}) : _logger = logger;
  final Logger _logger;

  bool _isScanning = false;
  WiFiNetwork? _currentNetwork;

  // Streams para notificaciones
  final StreamController<List<WiFiNetwork>> _networksController =
      StreamController<List<WiFiNetwork>>.broadcast();
  final StreamController<WiFiConnectionResult> _connectionController =
      StreamController<WiFiConnectionResult>.broadcast();

  /// Escanear redes WiFi disponibles
  Future<List<WiFiNetwork>> scanNetworks() async {
    if (_isScanning) {
      _logger.w('WiFi scan already in progress');
      return [];
    }

    try {
      _isScanning = true;

      // Solicitar permisos de ubicación (necesario para escanear WiFi en Android)
      final locationPermission = await Permission.location.request();
      if (!locationPermission.isGranted) {
        throw Exception('Location permission required for WiFi scanning');
      }

      final wifiScan = WiFiScan.instance;

      // Verificar si el escaneo esta soportado
      final canStartScan = await wifiScan.canStartScan();
      if (canStartScan != CanStartScan.yes) {
        _logger.w('WiFi scan not supported: $canStartScan');
        throw Exception('WiFi scanning is not supported on this device');
      }

      // Iniciar escaneo
      final startResult = await wifiScan.startScan();
      if (!startResult) {
        _logger.w('WiFi scan failed to start');
        throw Exception('Failed to start WiFi scan');
      }

      // Esperar a que el OS complete el escaneo
      await Future<void>.delayed(const Duration(seconds: 3));

      // Obtener resultados
      final canGetResults = await wifiScan.canGetScannedResults();
      if (canGetResults != CanGetScannedResults.yes) {
        _logger.w('Cannot get scan results: $canGetResults');
        throw Exception('Cannot retrieve WiFi scan results');
      }

      final accessPoints = await wifiScan.getScannedResults();

      // Mapear a WiFiNetwork, filtrar SSIDs vacios, 5GHz y deduplicar
      final networkMap = <String, WiFiNetwork>{};
      for (final ap in accessPoints) {
        if (ap.ssid.isEmpty) {
          continue;
        }
        // ESP32 solo soporta 2.4GHz (frecuencias < 3000 MHz)
        if (ap.frequency >= 3000) {
          continue;
        }
        final existing = networkMap[ap.ssid];
        if (existing == null || ap.level > existing.rssi) {
          networkMap[ap.ssid] = WiFiNetwork(
            ssid: ap.ssid,
            rssi: ap.level,
            security: ap.capabilities.contains('WPA3')
                ? 'WPA3'
                : ap.capabilities.contains('WPA2')
                    ? 'WPA2'
                    : ap.capabilities.contains('WPA')
                        ? 'WPA'
                        : ap.capabilities.contains('WEP')
                            ? 'WEP'
                            : 'Open',
            frequency: ap.frequency,
          );
        }
      }

      final networks = networkMap.values.toList()
        ..sort((a, b) => b.rssi.compareTo(a.rssi));

      _logger.i('Found ${networks.length} WiFi networks');
      _networksController.add(networks);
      return networks;
    } catch (e) {
      _logger.e('Error scanning WiFi networks: $e');
      rethrow;
    } finally {
      _isScanning = false;
    }
  }

  /// Conectar a una red WiFi
  Future<WiFiConnectionResult> connectToNetwork(
    WiFiCredentials credentials,
  ) async {
    try {
      // TODO(dev): Implement actual WiFi connection using a native plugin
      // This functionality requires system permissions and appropriate plugin
      _logger
        ..i('Attempting to connect to WiFi: ${credentials.ssid}')
        ..w('WiFi connection not implemented - requires native plugin');

      const result = WiFiConnectionResult(
        success: false,
        message:
            'WiFi connection not implemented. Use backend API to configure robot WiFi.',
      );

      _connectionController.add(result);

      throw UnimplementedError(
        'WiFi connection requires a native plugin or should be handled via backend API. '
        'For robot WiFi configuration, use RobotService.configureWiFi() instead.',
      );
    } on Exception catch (e) {
      _logger.e('Error connecting to WiFi: $e');
      final result = WiFiConnectionResult(
        success: false,
        message: 'Connection failed: $e',
      );

      _connectionController.add(result);
      return result;
    }
  }

  /// Desconectar de la red actual
  Future<void> disconnect() async {
    try {
      _logger.i('Disconnecting from WiFi');

      // Simular desconexión
      await Future<void>.delayed(const Duration(seconds: 1));

      _currentNetwork = null;

      const result = WiFiConnectionResult(
        success: true,
        message: 'Disconnected from WiFi',
      );

      _connectionController.add(result);
      _logger.i('WiFi disconnected');
    } catch (e) {
      _logger.e('Error disconnecting from WiFi: $e');
      rethrow;
    }
  }

  /// Obtener red actual conectada
  WiFiNetwork? get currentNetwork => _currentNetwork;

  /// Verificar si está conectado
  bool get isConnected => _currentNetwork != null;

  /// Verificar si está escaneando
  bool get isScanning => _isScanning;

  /// Obtener información de la red actual
  Map<String, dynamic>? get currentNetworkInfo {
    if (_currentNetwork == null) {
      return null;
    }

    return {
      'ssid': _currentNetwork!.ssid,
      'rssi': _currentNetwork!.rssi,
      'security': _currentNetwork!.security,
      'frequency': _currentNetwork!.frequency,
      'signalStrength': _getSignalStrength(_currentNetwork!.rssi),
      'isConnected': true,
    };
  }

  /// Calcular fuerza de señal basada en RSSI
  String _getSignalStrength(int rssi) {
    if (rssi >= -30) {
      return 'Excelente';
    }
    if (rssi >= -50) {
      return 'Buena';
    }
    if (rssi >= -70) {
      return 'Regular';
    }
    if (rssi >= -80) {
      return 'Débil';
    }
    return 'Muy débil';
  }

  /// Stream de redes encontradas
  Stream<List<WiFiNetwork>> get networksStream => _networksController.stream;

  /// Stream de resultados de conexión
  Stream<WiFiConnectionResult> get connectionStream =>
      _connectionController.stream;

  /// Verificar permisos necesarios
  Future<bool> checkPermissions() async {
    try {
      final locationPermission = await Permission.location.status;

      return locationPermission.isGranted;
    } on Exception catch (e) {
      _logger.e('Error checking permissions: $e');
      return false;
    }
  }

  /// Solicitar permisos necesarios
  Future<bool> requestPermissions() async {
    try {
      final locationStatus = await Permission.location.request();

      if (!locationStatus.isGranted) {
        _logger.w('Location permission was not granted');
      }

      return locationStatus.isGranted;
    } on Exception catch (e) {
      _logger.e('Error requesting permissions: $e');
      return false;
    }
  }

  /// Cerrar servicio
  Future<void> dispose() async {
    await _networksController.close();
    await _connectionController.close();
    _logger.i('WiFi Service disposed');
  }
}
